lib/spack/spack/package.py

```diff
@@ -1129,6 +1133,34 @@ def do_patch(self):
         else:
             touch(no_patches_file)
 
+    def patches_to_apply(self):
+        """If the patch set does not change between two invocations of spack,
+           then all patches in the set will be applied in the same order"""
+        patchesToApply = itertools.chain.from_iterable(
+            patch_list
+            for spec, patch_list in self.patches.iteritems()
+            if self.spec.satisfies(spec))
+        return sorted(patchesToApply, key=lambda p: p.path_or_url)
+
+    def content_hash(self, remove_pkg_name=False):
+        """Create a hash based on the sources and logic used to build the
+        package. This includes the contents of all applied patches and the
+        contents of applicable functions in the package subclass."""
+        hashContent = list()
+        # TODO? in cases where a digest or source_id isn't available, should
+        # this attempt to download the source and set one? For URL resources
+        # this will involve downloading the entire resource (repositories
+        # typically allow for just retrieving the latest commit ID)
```

Comment:
I like this idea in theory, but Spack currently assumes that the full spec hash is known before fetching.  I think we could engineer things so that it's not known until after fetching, but we'd need to rethink a number of assumptions to do that, and we'd need to bundle the fetches together before a build.  Those are all things that probably should happen, but not in this PR.

For now, if we can get away with it, ignore this issue.  Rationale (make sure this makes sense): the intent here is to build a content hash that we can use in a build farm, but *not* to identify builds (yet).  The place where this issue will come up is for `spack install foo@<version>` where `<version>` is not yet checksummed and known to Spack.  If we are not using the content hash in the DAG hash, then it shouldn't break spack if the content hashes for unknown versions are wrong.

Reply:
Reading through this code more, I think the only time this is false-y is when a source is a remote repository using a branch name like `master` rather than a tag or commit hash. Is that something we want to support?

Reply:
I changed it to output a warning and use an empty string for now.